Can't get node's public IP address from checkip.amazonaws.com

I followed instructions on https://github.com/rsksmart/rskj/wiki/install-rskj-using-fat-jar.

When I execute
`java -cp rskj-core-0.4.4-BAMBOO-all.jar co.rsk.Start`

I got following errors
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.springframework.cglib.core.ReflectUtils$1 (file:/Users/bhoomtawathplinsut/jars/rskj-core-0.4.4-BAMBOO-all.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of org.springframework.cglib.core.ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Exception in thread "main" org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'nodeManager' defined in URL [jar:file:/Users/bhoomtawathplinsut/jars/rskj-core-0.4.4-BAMBOO-all.jar!/org/ethereum/net/NodeManager.class]: Unsatisfied dependency expressed through constructor argument with index 0 of type [co.rsk.net.discovery.PeerExplorer]: Error creating bean with name 'peerExplorer' defined in org.ethereum.config.DefaultConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [co.rsk.net.discovery.PeerExplorer]: Factory method 'peerExplorer' threw exception; nested exception is java.lang.RuntimeException: Wildcard on bind address it's not allowed as fallback for public IP /0.0.0.0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'peerExplorer' defined in org.ethereum.config.DefaultConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [co.rsk.net.discovery.PeerExplorer]: Factory method 'peerExplorer' threw exception; nested exception is java.lang.RuntimeException: Wildcard on bind address it's not allowed as fallback for public IP /0.0.0.0
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:749)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:185)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1148)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1050)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:512)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:778)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:843)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:541)
	at co.rsk.SpringNodeBootstrapper.<init>(SpringNodeBootstrapper.java:43)
	at co.rsk.Start.main(Start.java:31)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'peerExplorer' defined in org.ethereum.config.DefaultConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [co.rsk.net.discovery.PeerExplorer]: Factory method 'peerExplorer' threw exception; nested exception is java.lang.RuntimeException: Wildcard on bind address it's not allowed as fallback for public IP /0.0.0.0
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:599)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1128)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1022)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:512)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.findAutowireCandidates(DefaultListableBeanFactory.java:1199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1123)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1021)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:814)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:741)
	... 14 more
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [co.rsk.net.discovery.PeerExplorer]: Factory method 'peerExplorer' threw exception; nested exception is java.lang.RuntimeException: Wildcard on bind address it's not allowed as fallback for public IP /0.0.0.0
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
	... 27 more
Caused by: java.lang.RuntimeException: Wildcard on bind address it's not allowed as fallback for public IP /0.0.0.0
	at org.ethereum.config.SystemProperties.getMyPublicIpFromRemoteService(SystemProperties.java:568)
	at org.ethereum.config.SystemProperties.getPublicIp(SystemProperties.java:540)
	at org.ethereum.config.DefaultConfig.peerExplorer(DefaultConfig.java:210)
	at org.ethereum.config.DefaultConfig$$EnhancerBySpringCGLIB$$befa31d5.CGLIB$peerExplorer$9(<generated>)
	at org.ethereum.config.DefaultConfig$$EnhancerBySpringCGLIB$$befa31d5$$FastClassBySpringCGLIB$$631f1e02.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:356)
	at org.ethereum.config.DefaultConfig$$EnhancerBySpringCGLIB$$befa31d5.peerExplorer(<generated>)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
	... 28 more
```

--- a/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java
+++ b/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java
@@ -38,6 +38,7 @@ import org.slf4j.LoggerFactory;
 
 import java.io.*;
 import java.net.InetAddress;
+import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.UnknownHostException;
 import java.nio.charset.StandardCharsets;
@@ -440,14 +441,15 @@ public abstract class SystemProperties {
 
     private String getMyPublicIpFromRemoteService(){
         try {
-            logger.info("Public IP wasn't set or resolved, using checkip.amazonaws.com to identify it...");
+            URL ipCheckService = publicIpCheckService();
+            logger.info("Public IP wasn't set or resolved, using {} to identify it...", ipCheckService);
 
-            try (BufferedReader in = new BufferedReader(new InputStreamReader(new URL("http://checkip.amazonaws.com").openStream()))) {
+            try (BufferedReader in = new BufferedReader(new InputStreamReader(ipCheckService.openStream()))) {
                 publicIp = in.readLine();
             }
 
             if (publicIp == null || publicIp.trim().isEmpty()) {
-                logger.warn("Unable to retrieve public IP from checkip.amazonaws.com {}.", publicIp);
+                logger.warn("Unable to retrieve public IP from {} {}.", ipCheckService, publicIp);
                 throw new IOException("Invalid address: '" + publicIp + "'");
             }
 
@@ -467,6 +469,10 @@ public abstract class SystemProperties {
         publicIp = bindAddress;
 
         return publicIp;
+    }
+
+    private URL publicIpCheckService() throws MalformedURLException {
+        return new URL(configFromFiles.getString("public.ipCheckService"));
     }
 
     public boolean isSyncEnabled() {

--- a/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java
+++ b/rskj-core/src/main/java/org/ethereum/config/SystemProperties.java
@@ -426,8 +426,6 @@ public abstract class SystemProperties {
                     publicIp = address.getHostAddress();
                     logger.info("Public IP identified {}", publicIp);
                     return publicIp;
-                } catch (IOException e) {
-                    logger.warn("Can't resolve public IP", e);
                 } catch (IllegalArgumentException e) {
                     logger.warn("Can't resolve public IP", e);
                 }
@@ -435,40 +433,40 @@ public abstract class SystemProperties {
             }
         }
 
-        publicIp = getMyPublicIpFromRemoteService();
+        publicIp = getMyPublicIpFromRemoteService().getHostAddress();
         return publicIp;
     }
 
-    private String getMyPublicIpFromRemoteService(){
+    private InetAddress getMyPublicIpFromRemoteService(){
         try {
             URL ipCheckService = publicIpCheckService();
             logger.info("Public IP wasn't set or resolved, using {} to identify it...", ipCheckService);
 
+            String ipFromService;
             try (BufferedReader in = new BufferedReader(new InputStreamReader(ipCheckService.openStream()))) {
-                publicIp = in.readLine();
+                ipFromService = in.readLine();
             }
 
-            if (publicIp == null || publicIp.trim().isEmpty()) {
-                logger.warn("Unable to retrieve public IP from {} {}.", ipCheckService, publicIp);
-                throw new IOException("Invalid address: '" + publicIp + "'");
+            if (ipFromService == null || ipFromService.trim().isEmpty()) {
+                logger.warn("Unable to retrieve public IP from {} {}.", ipCheckService, ipFromService);
+                throw new IOException("Invalid address: '" + ipFromService + "'");
             }
 
-            tryParseIpOrThrow(publicIp);
-            logger.info("Identified public IP: {}", publicIp);
-            return publicIp;
+            InetAddress resolvedIp = tryParseIpOrThrow(ipFromService);
+            logger.info("Identified public IP: {}", resolvedIp);
+            return resolvedIp;
         } catch (IOException e) {
             logger.error("Can't get public IP", e);
         } catch (IllegalArgumentException e) {
             logger.error("Can't get public IP", e);
         }
 
-        String bindAddress = getBindAddress().toString();
-        if (getBindAddress().isAnyLocalAddress()){
+        InetAddress bindAddress = getBindAddress();
+        if (bindAddress.isAnyLocalAddress()){
             throw new RuntimeException("Wildcard on bind address it's not allowed as fallback for public IP " + bindAddress);
         }
-        publicIp = bindAddress;
 
-        return publicIp;
+        return bindAddress;
     }
 
     private URL publicIpCheckService() throws MalformedURLException {
@@ -624,11 +622,16 @@ public abstract class SystemProperties {
         return configFromFiles.getString(PROPERTY_RPC_CORS);
     }
 
-    private InetAddress tryParseIpOrThrow(String ipToParse) throws IOException {
+    /**
+     * Parses a list of IPs separated by commas. E.g. "171.99.160.48, 171.99.160.48".
+     */
+    private InetAddress tryParseIpOrThrow(String ipsToParse) {
         try {
+            String[] ips = ipsToParse.split(", ");
+            String ipToParse = ips[ips.length - 1];
             return InetAddress.getByName(ipToParse);
         } catch (UnknownHostException e) {
-            throw new IllegalArgumentException("Invalid address: '" + ipToParse + "'", e);
+            throw new IllegalArgumentException("Invalid address(es): '" + ipsToParse + "'", e);
         }
     }
 

--- a/rskj-core/src/main/resources/reference.conf
+++ b/rskj-core/src/main/resources/reference.conf
@@ -123,7 +123,7 @@ public {
     # ip = google.com
 
     # The URL of a service returning our own public IP as the response body
-    ipCheckService = "http://checkip.amazonaws.com"
+    ipCheckService = "https://checkip.amazonaws.com"
 }
 
 # the number of blocks should pass before pending transaction is removed

--- a/rskj-core/src/main/resources/reference.conf
+++ b/rskj-core/src/main/resources/reference.conf
@@ -117,10 +117,14 @@ database {
 # Interface to bind peer discovery and wire protocol
 bind_address = 0.0.0.0
 
-# public IP/hostname which is reported as our host during discovery
-# if not set, the service http://checkip.amazonaws.com is used.
-# bind.address is the last resort for public ip when it cannot be obtained by other ways
-# public.ip = google.com
+public {
+    # public IP/hostname which is reported as our host during discovery.
+    # if not set, the ipCheckService URL is used.
+    # ip = google.com
+
+    # The URL of a service returning our own public IP as the response body
+    ipCheckService = "http://checkip.amazonaws.com"
+}
 
 # the number of blocks should pass before pending transaction is removed
 transaction.outdated.threshold = 10

--- a/rskj-core/src/main/resources/reference.conf
+++ b/rskj-core/src/main/resources/reference.conf
@@ -122,7 +122,8 @@ public {
     # if not set, the ipCheckService URL is used.
     # ip = google.com
 
-    # The URL of a service returning our own public IP as the response body
+    # The URL of a service returning our own public IP as the response body.
+    # If multiple IPs are returned separated by commas (e.g Amazon returns X-Forwarded-For values) the last one is used
     ipCheckService = "https://checkip.amazonaws.com"
 }
 

